### ISSUE TITLE
class id ignore clarification in IOU metrics

### ISSUE THREAD

**Issue by pure-rgb** [keras-team/keras#21907]
in the following test case, y true contains (0, 1, 2, 255) total num_class is 4. As 255 id will be ignored, the total class id thereby became 3, or it should be kept 4 and during loss calculation the ignore id will mask out the 255 indices?

https://github.com/keras-team/keras/blob/f0a48a6f3ff807eeecc0e00f4417c7e099d947b9/keras/src/metrics/iou_metrics_test.py#L250C5-L266C64

```python
def test_unweighted_ignore_class_255(self):
        y_pred = [0, 1, 1, 1]
        y_true = [0, 1, 2, 255]

        m_obj = metrics.MeanIoU(num_classes=3, ignore_class=255)

        result = m_obj(y_true, y_pred)

        # cm = [[1, 0, 0],
        #       [0, 1, 0],
        #       [0, 1, 0]]
        # sum_row = [1, 1, 1], sum_col = [1, 2, 0], true_positives = [1, 1, 0]
        # iou = true_positives / (sum_row + sum_col - true_positives))
        expected_result = (
            1 / (1 + 1 - 1) + 1 / (2 + 1 - 1) + 0 / (0 + 1 - 0)
        ) / 3
        self.assertAllClose(result, expected_result, atol=1e-3)
```

what if i want to ignore class id 2. and what if i want to ignore class id 2 and 255 at the same time

**Comment by maitry63** [comment]
Hi @pure-rgb , num_classes never changes
- The num_classes argument in Mean_IoU sets the size of the confusion matrix .
- Ignored classes do not reduce this number.

ignore_class only masks pixels.
- Pixels in y_true equal to ignore_class are skipped when building the confusion matrix.
- Other classes are computed normally.

Multiple ignored classes.
- keras MeanIoU supports only one ignore_class.
- To ignore multiple labels, remap all of them to a single sentinel value (eg,-1) and pass that as ignore_class.

IoU averaging.
- The final mean IoU is computed as the average over all num_classes.
- Ignored classes contribute 0 to the numerator, but are effectively masked out in calculations if using ignore_class. 

Check this [Documentation](https://keras.io/api/metrics/segmentation_metrics) for details.

**Comment by pure-rgb** [comment]
@maitry63 

> Hi @pure-rgb , num_classes never changes
The num_classes argument in Mean_IoU sets the size of the confusion matrix .
Ignored classes do not reduce this number.

Its confusing due to how the implementation is done. Shouldn't the implementation do like below:

```python
y_pred = [0, 1, 1, 1]
y_true = [0, 1, 2, 255]

m_obj = metrics.MeanIoU(num_classes=4, ignore_class=255)

result = m_obj(y_true, y_pred)

# cm = [[1, 0, 0],
#       [0, 1, 0],
#       [0, 1, 0]]
#       [0, 0, 0]]
```

**Comment by lucasdavid** [comment]
255 (or -1) is usually a code associated with `void`, and not generally considered a class. Hence the test you linked has `num_classes=3`.

> what if i want to ignore class id 2. 

You set `metrics.MeanIoU(num_classes=3, ignore_class=2)`

> and what if i want to ignore class id 2 and 255 at the same time

Unfortunately, this use-case is not uncovered, as the original idea was to be an exact replica of torch's `ignore_index` feature and because an arbitrary number of ignored elements would require outer-broadcasted comparison for tensorflow (`reshape(reduce_any(reshape(y_true, (-1, 1)) == reshape(ignored_elements, (1, -1)), axis=1), y_true.shape)`).

If you want this, you will need to mask these values yourself:
```py

m_obj = metrics.MeanIoU(num_classes=2, ignore_class=255)

for step, (x, y, sample_weight) in enumerate(dataset):
  y_pred = [0, 1, 1, 1]  # model(x, training=True)
  y_true = [0, 1, 2, 255]  # y
  y_true = tf.where((y_true != 2) & (y_true != 255), y_true, 255)

  m_obj.update_state(y_true, y_pred, sample_weight=sample_weight)
```

Note: a larger `num_classes` or masking intermediate values (i.e., y_true=[0, 1, 2, 3] and ignore_class=1) should be fine due to [iou_metrics#L291](https://github.com/keras-team/keras/blob/adbfd13426a0da9864d9a0fcd5be5eed74ca341f/keras/src/metrics/iou_metrics.py#L291).

**Comment by maitry63** [comment]
Hi @pure-rgb , In `metrics.MeanIoU`, the num_classes argument defines the shape of the confusion matrix,  it is not automatically reduced when you specify `ignore_class`. This option simply excludes samples with that label from the metric computation, so it doesn’t change the number of classes used to allocate the confusion matrix.
In your case -
```
y_pred = [0, 1, 1, 1]
y_true = [0, 1, 2, 255]
m_obj = metrics.MeanIoU(num_classes=3, ignore_class=255)
```
the metric still uses a confusion matrix for classes 0,1,2. The sample with class 255 is masked out , hence it doesn’t affect the counts, but num_classes stays 3. That’s why the expected IoU calculation in the test makes sense. 

Also note that ignore_class only accepts a single class ID. If you want to ignore multiple IDs (e.g., both 2 and 255), you’ll need to filter them out before passing the data to the metric. 
Let me know if this helps!
